contracts/dev-utils/contracts/src/OrderValidationUtils.sol

```diff
@@ -73,36 +73,65 @@ contract OrderValidationUtils is
         // and the total amounts specified in the order
         uint256 transferableTakerAssetAmount;
-        if (order.makerAssetData.equals(order.makerFeeAssetData)) {
-            // If `makerAsset` equals `makerFeeAsset`, the % that can be filled is
-            // transferableMakerAssetAmount / (makerAssetAmount + makerFee)
+
+        // If `makerFee` is 0, the % that can be filled is (transferableMakerAssetAmount / makerAssetAmount)
+        if (makerFee == 0) {
             transferableTakerAssetAmount = LibMath.getPartialAmountFloor(
                 transferableMakerAssetAmount,
-                order.makerAssetAmount.safeAdd(makerFee),
+                makerAssetAmount,
                 takerAssetAmount
             );
         } else {
-            // Get the transferable amount of the `makerFeeAsset`
-            uint256 transferableMakerFeeAssetAmount = getTransferableAssetAmount(makerAddress, order.makerFeeAssetData);
-
-            // If `makerFee` is 0, the % that can be filled is (transferableMakerAssetAmount / makerAssetAmount)
-            if (makerFee == 0) {
+            if (order.makerFeeAssetData.equals(order.makerAssetData)) {
+                // The % that can be filled is transferableMakerAssetAmount / (makerAssetAmount + makerFee)
                 transferableTakerAssetAmount = LibMath.getPartialAmountFloor(
                     transferableMakerAssetAmount,
-                    order.makerAssetAmount,
+                    makerAssetAmount.safeAdd(makerFee),
                     takerAssetAmount
                 );
-
-            // If `makerAsset` does not equal `makerFeeAsset`, the % that can be filled is the lower of
-            // (transferableMakerAssetAmount / makerAssetAmount) and (transferableMakerAssetFeeAmount / makerFee)
+            } else if (order.makerFeeAssetData.equals(order.takerAssetData)) {
+                // The % that can be filled is
+                // (transferableMakerAssetAmount + takerAssetAmount) / (makerAssetAmount + makerFee)
+                transferableTakerAssetAmount = LibMath.getPartialAmountFloor(
+                    transferableMakerAssetAmount.safeAdd(takerAssetAmount),
+                    makerAssetAmount.safeAdd(makerFee),
+                    takerAssetAmount
+                );
+            } else if (
+                order.takerAssetData.readBytes4(0) == IAssetData(address(0)).MultiAsset.selector
+                && multiAssetContainsAsset(order.takerAssetData, order.makerFeeAssetData)
+            ) {
+                // find the asset data in order.takerAssetData which matches the makerFeeAsset
+                (
+                    ,
+                    uint256[] memory amounts,
+                    bytes[] memory nestedAssetData
+                ) = decodeMultiAssetData(order.takerAssetData);
+
+                for (uint i=0; i < amounts.length; i++) {
+                    if (order.makerFeeAssetData.equals(nestedAssetData[i])) {
```

Comment:
Let's create a helper function that compares the first 4 bytes of each `assetData`. We do this a lot throughout these contracts.